pkg/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,370 @@
+func (c *CacheManager) ExcludeProcesses(newExcluder *process.Excluder) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.processExcluder.Equals(newExcluder) {
+		return
+	}
+
+	c.processExcluder.Replace(newExcluder)
+	// there is a new excluder which means we need to schedule a wipe for any
+	// previously watched GVKs to be re-added to get a chance to be evaluated
+	// for this new process excluder.
+	c.excluderChanged = true
+}
+
+func (c *CacheManager) AddObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	gvk := instance.GroupVersionKind()
+
+	isNamespaceExcluded, err := c.processExcluder.IsNamespaceExcluded(process.Sync, instance)
+	if err != nil {
+		return fmt.Errorf("error while excluding namespaces for gvk: %+v: %w", gvk, err)
+	}
+
+	// bail because it means we should not be
+	// syncing this gvk's objects as it is namespace excluded.
+	if isNamespaceExcluded {
+		c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+		return nil
+	}
+
+	syncKey := syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName())
+	if c.watchedSet.Contains(gvk) {
+		_, err = c.opa.AddData(ctx, instance)
+		if err != nil {
+			c.syncMetricsCache.AddObject(
+				syncKey,
+				syncutil.Tags{
+					Kind:   instance.GetKind(),
+					Status: metrics.ErrorStatus,
+				},
+			)
+
+			return err
+		}
+	}
+
+	c.tracker.ForData(instance.GroupVersionKind()).Observe(instance)
+
+	c.syncMetricsCache.AddObject(syncKey, syncutil.Tags{
+		Kind:   instance.GetKind(),
+		Status: metrics.ActiveStatus,
+	})
+	c.syncMetricsCache.AddKind(instance.GetKind())
+
+	return err
+}
+
+func (c *CacheManager) RemoveObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	gvk := instance.GroupVersionKind()
+
+	if c.watchedSet.Contains(gvk) {
+		if _, err := c.opa.RemoveData(ctx, instance); err != nil {
+			return err
+		}
+	}
+
+	// only delete from metrics map if the data removal was succcesful
+	c.syncMetricsCache.DeleteObject(syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName()))
+	c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+
+	return nil
+}
+
+func (c *CacheManager) wipeData(ctx context.Context) error {
+	if _, err := c.opa.RemoveData(ctx, target.WipeData()); err != nil {
+		return err
+	}
+
+	// reset sync cache before sending the metric
+	c.syncMetricsCache.ResetCache()
+	c.syncMetricsCache.ReportSync()
+
+	return nil
+}
+
+func (c *CacheManager) ReportSyncMetrics() {
+	c.syncMetricsCache.ReportSync()
+}
+
+func (c *CacheManager) syncGVK(ctx context.Context, gvk schema.GroupVersionKind) error {
+	u := &unstructured.UnstructuredList{}
+	u.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind + "List",
+	})
+
+	var err error
+	c.mu.Lock()
+	if !c.watchedSet.Contains(gvk) {
+		// we are not actually watching this gvk anymore
+		// so don't list instances for it.
+		err = nil
```

Comment:
Better to just early return since we won't list and won't add any objects.

Reply:
+1, but also we should wrap this in an anonymous function call:

func() {
  c.mu.Lock()
  defer c.mu.Unlock()
  
  ... do some things
} ()

acquiring a mutex w/o a defer risks deadlocking (which adding a return would do in this case.